### PR TITLE
[WIP] Extended YAML support

### PR DESCRIPTION
**Pull request by cgranade**
This PR (still work in progress) adds to existing YAML config-file support by updating it for Py3 and by allowing config files to specify attributes to be set on newly-created instruments. This is useful, for instance, in the case of ThorLabs APT motor controller devices, where the model of each attached motor must be manually specified. Since this is a fairly significant set of changes, I wanted to open the PR in advance of writing tests and better documentation so as to get feedback on the design.

In particular, this PR adds the following functionality:

- A new special ``test://`` instrument URI schema to make it easier to test config support in lieu of actual devices.
- Support for instantiating unitful numbers from within YAML files, with a new ``!Q`` tag.
- Support for new ``attrs`` config sections of the following form:
```yaml
instrument_id:
    class: !!python/name:instruments.ClassName
    uri: schema://inst
    attrs:
        foo: bar
        baz[0].blah: !Q 1 tesla
```
- Slight refactoring to the ThorLabs APT classes to allow for better use with config files (backwards-compatible, guarded with a ``DeprecationWarning``).

I intend on adding the following to the PR:
- [ ] Removing all pylint warnings introduced.
- [ ] Writing tests for the new attr minilanguage, YAML tags, ``test://`` URI, and ``attrs`` config sections.
- [ ] Adding references to new functionality to the user guide.

### CODE REVIEW

File: instruments/config.py [hunk 1]
```diff
@@ -98,20 +139,30 @@ def load_instruments(conf_file_name, conf_path="/"):
...
+        with open(conf_file_name, 'r') as f:
+            conf_dict = yaml.load(f)
+    else:
+        conf_dict = yaml.load(conf_file_name)
 
     conf_dict = walk_dict(conf_dict, conf_path)
 
     inst_dict = {}
-    for name, value in conf_dict.iteritems():
+    for name, value in conf_dict.items():
```
Comment (scasagrande) [inline instruments/config.py]: good catch

File: instruments/config.py [hunk 2]
```diff
@@ -63,6 +83,27 @@ def load_instruments(conf_file_name, conf_path="/"):
     the form
     ``{'ddg': instruments.srs.SRSDG645.open_from_uri('gpib+usb://COM7/15')}``.
 
+    Optionally, each the value of each instrument key can also contain a dictionary
```
Comment (scasagrande) [inline instruments/config.py]: awkward sentence

File: instruments/config.py [hunk 3]
```diff
@@ -63,6 +83,27 @@ def load_instruments(conf_file_name, conf_path="/"):
...
+    of attributes to set on the newly-created instrument. For instance, the following
+    dictionary creates a ThorLabs APT motor controller instrument with a single motor
+    model configured::
+
+        rot_stage:
+            class: !!python/name:instruments.thorabsapt.APTMotorController
+            uri: serial:///dev/ttyUSB0?baud=115200
+            attrs:
```
Comment (scasagrande) [inline instruments/config.py]: I LOVE IT

File: instruments/tests/test_config.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from instruments import Instrument
+from instruments.config import (
+    load_instruments
+)
+
+# TEST CASES #################################################################
+
+# pylint: disable=protected-access,missing-docstring
+
+@skipIf(yaml is None, "PyYAML is not installed.")
```
Comment (scasagrande) [inline instruments/tests/test_config.py]: I don't think we should be skipping tests if pyyaml is missing. I can see the situation where it fails to install in CI (for whatever reason) and simultaneously there is a bug that these tests would reveal. Build would be green, even though some tests technically fail.

File: instruments/config.py (hunk 2 above)
Comment (cgranade) [inline instruments/config.py]: Good point, I'll improve that, then.

File: instruments/tests/test_config.py (hunk 4 above)
Comment (cgranade) [inline instruments/tests/test_config.py]: Fair point, I'll make sure YAML (whichever once we go with) is appropriately a dependency for the test environment, then. I think now that the ``conda`` command depends on ``rumel.yaml``, it's not as heavy of a dependency as it used to be.

**Comment by cgranade** [comment]
Thanks, I'm glad you like! Thanks as well for the feedback, I'll go on and fix that stuff up, then. (Sorry about the wrong target for the PR, I always forget to change from the GitHub default... have they introduced a setting to control that yet?)

**Pull request closed without merging**